servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/FilterableClientToClient.java

```diff
@@ -196,4 +203,65 @@ public Completable closeAsyncGracefully() {
     public StreamingHttpRequest newRequest(final HttpRequestMethod method, final String requestTarget) {
         return client.newRequest(method, requestTarget);
     }
+
+    private static Single<StreamingHttpResponse> executeRequest(final StreamingHttpRequester requester,
+                                                                final StreamingHttpRequest request,
+                                                                final HttpExecutionStrategy strategy,
+                                                                final Object lock) {
+        return Single.defer(() -> {
+            final ContextMap context = request.context();
+
+            if (SKIP_CONCURRENT_REQUEST_CHECK) {
+                return executeRequest(requester, request, context, strategy).shareContextOnSubscribe();
+            }
+
+            // Prevent concurrent execution of the same request through the same layer.
+            // In general, we do not expect users to execute the same mutable request concurrently. Therefore, the cost
+            // of synchronized block should be negligible for most requests, unless they messed up with reactive streams
+            // chain and accidentally subscribed to the same request concurrently. This protection helps them avoid
+            // ambiguous runtime behavior caused by a corrupted mutable request state.
+            final Object inFlight;
+            // Note that because request.context() lazily allocates a new ContextMap, there is a risk that
+            // synchronization will happen on two different contexts. However, this is acceptable compromise because:
+            //  - Most likely users subscribe 2+ times to the same request from the same thread.
+            //  - This is the best effort protection, giving users at least one rejection should be enough to let them
+            //    know their code is incorrect and should be rewritten.
```

Comment:
Would it be worth it to elevate the synchronization to the request itself?

Reply:
Because of gRPC meta-data we should synchronized on the context. Otherwise, underlying HTTP requests will be different.